arthur/tasks.py

```diff
@@ -204,13 +208,16 @@ def remove(self, task_id):
         :raises NotFoundError: raised when the given task identifier
             is not found on the registry
         """
-        try:
-            self._rwlock.writer_acquire()
-            del self._tasks[task_id]
-        except KeyError:
+        self._rwlock.reader_acquire()
+        found = self.conn.exists(task_id)
+        self._rwlock.reader_release()
+
+        if not found:
             raise NotFoundError(element=str(task_id))
-        finally:
-            self._rwlock.writer_release()
+
+        self._rwlock.writer_acquire()
+        self.conn.expire(task_id, -1)
+        self._rwlock.writer_release()
```

Comment:
This is not a race condition because the way redis works which is that if the key doesn't exist it would do nothing. The problem is you need to know exactly the behaviour of `expire`.

I think it's better to follow a defensive programming, which will be to include everything under a writer block.

Reply:
The same happens in other parts of the code.